src/qat/purr/integrations/qasm.py

```diff
@@ -944,16 +737,335 @@ def __repr__(self):
         return f"Partial instance of {type(self._ref_instance).__name__}"
 
 
+class Qasm3ParserBase(AbstractParser, QASMVisitor):
+
+    def parser_language(self) -> Languages:
+        return Languages.Qasm3
+
+    def _includes_standard_gates(self, qasm_str: str) -> bool:
+        return any(
+            [
+                include in qasm_str
+                for include in ('include "qelib1.inc";', 'include "stdgates.inc";')
+            ]
+        )
+
+    def load_default_gates(self, context) -> QasmContext:
+        node = ast.Include(filename="stdgates.inc")
+        self.visit(node, context)
+        return context
+
+    def parse(self, builder, qasm: str) -> InstructionBuilder:
+        self.builder = builder
+        # Parse or pick up the cached version, then remove it as we're about to
+        # interpret it.
+        program = self._fetch_or_parse(qasm)
+        if (qasm_id := hash(qasm)) in self._cached_parses:
+            del self._cached_parses[qasm_id]
+        context = QasmContext()
+        if not self._includes_standard_gates(qasm_str):
```

Comment:
I'm a bit confused here, why are gates only loaded if there isn't an include in the header? Could be misreading here

Reply:
Basically to not waste time loading the gates twice.